definitions/marts/looker_studio_marts/ecf/ls_ecf_provider_activity_schedule_changes.sqlx

```diff
@@ -0,0 +1,117 @@
+	  user_id: "The user ID taken from the teacher profile associated with the participant profile.",
+	  participant_type: "The type of the participant whose schedule has been changed. The value can be either ECT or Mentor.",
+	  request_path: "The full API path used by the lead provider to change the participant's schedule. The path contains the user_id but always ends with 'change-schedule'.",
+	  key_updated: "The name of the field in the table participant_profiles_latest_cpd that was changed by the API request. This is always schedule_id as that is the only field a lead provider can amend via the 'change-schedule' API path.",
+	  new_schedule: "The new value for the schedule_id field on the participant profile translated into text via schedules_latest_cpd.",
+	  new_cohort: "The cohort associated with the new schedule. Even if the cohort has not changed this field will be populated with the cohort that is associated with the new schedule.",
+	  previous_schedule: "The previous value for the schedule_id field on the participant profile translated into text via schedules_latest_cpd.",
+	  previous_cohort: "The cohort associated with the previous schedule. Even if the cohort has not changed this field will be populated with the cohort that is associated with the previous schedule."
+    }
+}
+
+-- Gather ECF Induction Records to be able to join User ID, Participant Type & Cohort onto participants within the script without needing to join on Participant Profiles and Teacher Profiles separately.
+WITH ecf_inductions AS(
+  SELECT
+     induction_record_id
+    ,participant_profile_id
+    ,participant_type
+    ,user_id
+    ,cohort
+    ,DATE(start_date) AS start_date
+    ,IFNULL(DATE(end_date), DATE(2050,12,31)) AS end_date
+  FROM
+    ${ref('ecf_inductions')}
+)
+
+-- Gather Schedule Changes made via the API by Lead Providers. These API changes impact the participant profile and as such we take the information from the field changes made to this table. When a schedule changes so can the cohort and in this CTE we capture both values in their new and previous states.
+SELECT 
+   DATE(occurred_at) AS event_date
+  ,DATE_DIFF(CURRENT_DATE(), DATE(occurred_at), DAY) AS event_date_delta
+  ,lead_providers.name AS cpd_lead_provider_name
+  ,request_uuid
+  ,entity_id AS participant_profile_id
+  ,ecf_inductions.user_id
+  ,ecf_inductions.participant_type AS participant_type
+  ,request_path 
+  ,key_updated
+  ,new_schedules.schedule_name AS new_schedule
+  ,new_schedules.cohort AS new_cohort
+  ,prev_schedules.schedule_name AS previous_schedule
+  ,prev_schedules.cohort AS previous_cohort
+FROM 
+ ${ref('participant_profiles_field_updates_cpd')} AS profiles
+-- Join on ecf_inductions CTE and de-duplicate where multiple induction records are valid around the date of the event. We use the last record valid on the date.
+LEFT JOIN
+  (
+    SELECT
+        *
+      ,ROW_NUMBER() OVER(PARTITION BY participant_profile_id ORDER BY end_date DESC) AS rn0

```

Comment:
Why isn't this happening in the CTE above?

Reply:
The RN needs the date onto which it is being joined in order to correctly determine position 1. If the RN is calculated in the previous CTE duplicates appear when joined.

Reply:
But you should be able to do that in the CTE above, you can perform the ifnull() in the window statement and then run a qualify?

Reply:
May need to discuss this one. When I do a RN on the first CTE the record with 1 is the latest. If I include the dates in the partition by the record with 1 is the last record per start & end date grouping. When joining on an event date between start & end there are more than 1 induction records being joined and I need the one that was most relevant on the date. Happy to have a call on this as I haven't found a way to do it in the first CTE that works well in the second.